README
==========================
ARandR: Another XRandR GUI
==========================


ARandR is designed to provide a simple visual front end for XRandR_. Relative monitor positions are shown graphically and can be changed in a drag-and-drop way.



Features
--------

* Full control over output positioning (instead of plain "left of") with edge snapping
* Saving configurations as executable shell scripts (configurations can be loaded without using this program)
* Configuration files can be edited to include additional payload (like xsetwacom_ commands tablet PC users need when rotating), which is preserved when editing
* Main widget separated from packaged application (to facilitate integration with existing solutions)



Installation
------------

* Using your distribution's installer:

  * Debian_ / Ubuntu_::

      aptitude install arandr

  * archlinux_::

      pacman -S arandr

  * Gentoo_::

      emerge arandr

  * Fedora_::

      yum install arandr

  * OpenSUSE_::

      zypper in arandr

  * Slackware: see slackbuilds.org_

* From source:

  * Fetch either

    * the `latest release`_::

        wget http://christian.amsuess.com/tools/arandr/files/arandr-0.1.10.tar.gz
        tar xzf arandr-0.1.10.tar.gz
        cd arandr-0.1.10

    * or get it directly from GIT (`hosted on GitLab`_)::

        git clone https://gitlab.com/arandr/arandr
        cd arandr

  * and start it using::

      ./arandr

  * or install it::

      sudo ./setup.py install



Dependencies
------------

* python_ 3
* PyGObject_
* xrandr_
* docutils_ (for building the man page)



Bugs / Communication
--------------------

* There exists a `list of bugs`_ (and feature requests and general wishlist). If there is a problem or you want to request a feature, just `report it`_.
* New releases are announced as published tags_ and can be subscribed as an `ATOM feed`_.

Ongoing development
-------------------

An overhauled version of ARandR is in the works in the `next` branch found in the git repository. Feature parity with the current releases has not been reached, but some interesting features are already implemented:

* Parsing of the complete `--verbose` output, including properties
* Setting refresh rates
* Auto-discovery of DISPLAY in remote contexts 
* Simulation and regression test infrastructure

Features targetted for the first release from the new branch:

* Setting properties
* Parsing EDID information


Similar projects
----------------

A number of other programs exist that cover similar functionality, often
bundled with a particular desktop environment. Those I know of are, in
alphabetical order:

* Gnome's ``gnome-control-center display``, notable for its ``~/.config/monitors.xml`` `persistence <https://wiki.gnome.org/Attic/RandR#Storage_of_RANDR_configurations>`_
* `grandr <ftp://ftp.freedesktop.org/pub/individual/app/>`_
* `grandr-applet <https://sites.google.com/site/kdekorte2/grandr_applet>`_ (which seems not to be related to grandr)
* KDE's RandR tool ``krandrtray``, notable for being the only listed program to be Qt based
* LXDE's `LXRandR <http://wiki.lxde.org/en/LXRandR>`_, notable for storing configuration in ``~/.config/autostart/lxrandr-autostart``
* `urandr <http://www.albertomilone.com/wordpress/?p=288>`_
* `zarfy <http://sourceforge.net/projects/zarfy/>`_, notable for its nice use of a current screenshot



Contributing
------------

If you are fluent with languages not yet supported, you can add translations to ARandR using gettext_. An easy way to do this is `ARandR's page on hosted Weblate`_.

If you want to contribute code, contact me_ directly or send bug reports, suggestions, patches and git pull requests to the bugtracker_.

If you want to support the project via flattr, you can `flattr this`_.


About
-----

Copyright © chrysn_ <chrysn@fsfe.org> 2008–2019, Себастьян Gli ţa Κατινα 2011, Johannes Holmberg <johannes@update.uu.se> 2015, actionless 2019 published under GPLv3_ or any later version.

Inspired by the `dual head sketch`_ in the ThinkWiki_.

Translations by

* Ajeje Brazorf (Sardinian)
* Alex Kalles, Dimitris Giouroukis, Efstathios Iosifidis (Modern Greek (1453-))
* Algimantas Margevičius, Mantas Kriaučiūnas, Moo (Lithuanian)
* Alireza Savand, reza khan (Persian)
* Allan Nordhøy (Norwegian Bokmål (Norway))
* Anders Jonsson, Erik Sundin, Ingemar Karlsson (Swedish)
* Bakr Al-Tamimi, Mohammad Alhargan, aboodilankaboot (Arabic)
* Balázs Úr, Tamás Nagy (Hungarian)
* Belvar, tornoz (Breton)
* Bruno_Patri, Clément Démoulins, Tuux (French)
* Calin Sopterean, Себастьян Gli ţa Κατινα (Romanian)
* Carezero, ChuChangMing, ykla (Chinese (China))
* DARREN MELROY MENEZES, gundachandru (Kannada)
* Denis Jukni (Albanian)
* Fred Maranhão, Guilherme Souza Silva, Phantom X (Portuguese (Brazil))
* Fuxik Nah, Slavko (Slovak)
* HsH, Ivan Vantu5z, Vladimir (Russian)
* Igor, Olexandr Nesterenko, Rax Garfield (Ukrainian)
* Joe Hansen, Sebastian Wahl (Danish)
* Kristjan Räts (Estonian)
* Louies (Chinese (Hant))
* Lu Ca (Portuguese (Brazil), Sardinian)
* Luca Vetturi, Quizzlo (Italian)
* Luis García Sevillano, Ricardo A. Hermosilla Carrillo (Spanish)
* Mehmet Gülmen, cem guresci (Turkish)
* Michal Čihař (Czech)
* Miguel A. Bouzada (Galician)
* Miguel Anxo Bouzada (Galician, Spanish)
* Nobuhiro Iwamatsu, o-157 (Japanese)
* ParkJS, cho bkwon, pCsOrI (Korean (Korea, Republic of))
* Piotr Strebski, RooTer, agilob (Polish)
* Rasti K5 (Sorani)
* Semsudin Abdic (Bosnian)
* Slobodan Simić (Serbian)
* Yaron Shahrabani (Hebrew)
* chrysn (English, German)
* ditokp (Indonesian)
* ekeimaja (Finnish)
* el_libre, josep constanti (Catalan)
* ikmaak, wimfeijen (Dutch)


 .. _XRandR: http://www.x.org/wiki/Projects/XRandR
 .. _xsetwacom: http://linuxwacom.sourceforge.net/index.php/howto/xsetwacom
 .. _`latest release`: http://christian.amsuess.com/tools/arandr/files/arandr-0.1.9.tar.gz
 .. _archlinux: http://www.archlinux.org/packages/community/any/arandr/
 .. _chrysn: http://christian.amsuess.com
 .. _GPLv3: http://www.gnu.org/licenses/gpl-3.0.txt
 .. _`dual head sketch`: http://www.thinkwiki.org/wiki/Image:Intel-DualHead.png
 .. _ThinkWiki: http://thinkwiki.org/
 .. _`hosted on GitLab`: https://gitlab.com/arandr/arandr
 .. _Debian: http://packages.debian.org/arandr
 .. _Ubuntu: http://packages.ubuntu.com/arandr
 .. _gettext: http://www.gnu.org/software/gettext/
 .. _`ARandR's page on hosted Weblate`: http://hosted.weblate.org/projects/arandr/translations/
 .. _me: mailto:chrysn@fsfe.org
 .. _bugtracker: https://gitlab.com/arandr/arandr/issues
 .. _python: http://www.python.org/
 .. _PyGObject: https://pygobject.readthedocs.io/en/latest/
 .. _slackbuilds.org: https://slackbuilds.org/repository/14.1/desktop/arandr/
 .. _Gentoo: http://packages.gentoo.org/package/x11-misc/arandr
 .. _Fedora: https://apps.fedoraproject.org/packages/arandr
 .. _OpenSUSE: https://software.opensuse.org/package/arandr
 .. _docutils: http://docutils.sourceforge.net/
 .. _`debian bug #507521`: http://bugs.debian.org/507521
 .. _`X11:Utilities repository`: http://download.opensuse.org/repositories/X11:/Utilities/
 .. _`list of bugs`: https://gitlab.com/arandr/arandr/issues
 .. _`report it`: https://gitlab.com/arandr/arandr/issues/new
 .. _`flattr this`: https://flattr.com/submit/auto?user_id=chrysn&url=http%3A%2F%2Fchristian.amsuess.com%2Ftools%2Farandr%2F
 .. _tags: https://gitlab.com/arandr/arandr/tags
 .. _`ATOM feed`: https://gitlab.com/arandr/arandr/tags?format=atom


debian/changelog
arandr (0.1.9-2+rpt7) buster; urgency=medium

  * Modify KMS check to include full KMS on Pi 4

 -- Simon Long <simon@raspberrypi.com>  Tue, 02 Jun 2020 10:42:08 +0100

arandr (0.1.9-2+rpt6) buster; urgency=medium

  * Separate out refresh rate into separate menu

 -- Simon Long <simon@raspberrypi.org>  Thu, 11 Jul 2019 09:45:44 +0100

arandr (0.1.9-2+rpt5) buster; urgency=medium

  * Fix file close bug

 -- Simon Long <simon@raspberrypi.org>  Mon, 01 Jul 2019 09:22:34 +0100

arandr (0.1.9-2+rpt4) unstable; urgency=medium

  * Sanity check on saved config before loading on boot
  * Allow setting of display refresh rate

 -- Simon Long <simon@raspberrypi.org>  Fri, 21 Jun 2019 12:34:31 +0100

arandr (0.1.9-2+rpt3) unstable; urgency=medium

  * Fix revert behaviour; escape key now reverts

 -- Simon Long <simon@raspberrypi.org>  Thu, 16 May 2019 07:43:54 +0100

arandr (0.1.9-2+rpt2) unstable; urgency=medium

  * Only run xrandr commands if FKMS is in use

 -- Simon Long <simon@raspberrypi.org>  Tue, 14 May 2019 14:12:42 +0100

arandr (0.1.9-2+rpt1) unstable; urgency=medium

  * Various UI tweaks

 -- Simon Long <simon@raspberrypi.org>  Wed, 01 May 2019 13:01:20 +0100

arandr (0.1.9-2) unstable; urgency=medium

  * Change debhelper compat to 10
  * Update standards-version to 3.9.8
  * Explicitly enable dh_python2 (Closes: #849689)
  * Drop needless debian/dirs
  * Copyright: Use https URIs

 -- Christian M. Amsüss <chrysn@fsfe.org>  Wed, 25 Jan 2017 11:38:24 +0100

arandr (0.1.9-1) unstable; urgency=medium

  * New upstream release
    - Declare compatibility with xrandr 0.1.5 (Closes: #815731)
    - New translations
  * Copyright file updates
  * Drop .menu file (Closes: #797973)
  * Change Standards-Version to 3.9.7 (no changes)
  * Switch to secure Vcs- URIs

 -- Christian M. Amsüss <chrysn@fsfe.org>  Wed, 24 Feb 2016 14:00:35 +0100

arandr (0.1.8-1) unstable; urgency=medium

  * New upstream release
    - Allow setting the primary output
    - New translations
  * Change Standards-Version to 3.9.6 (no changes)
  * Update package source addresses
  * Replace shipped license file with symlink to common-licenses
  * Copyright file updates

 -- Christian M. Amsüss <chrysn@fsfe.org>  Fri, 26 Jun 2015 12:36:06 +0200

arandr (0.1.7.1-1) unstable; urgency=low

  * New upstream release
    - Compatibility with non "<width>x<height" modes (Closes: #507521)
    - Compatibility with xrandr client 1.4.0 ("primary" issue)
    - New translations
  * Note new translations and contributor in copyright file
  * Change Standards-Version to 3.9.4 (no changes)
    - Specify branch of Vcs-git field
  * Add a Debian menu file (Closes: #692176)

 -- Christian M. Amsüss <chrysn@fsfe.org>  Mon, 04 Mar 2013 19:11:30 +0100

arandr (0.1.6-1) unstable; urgency=low

  * New upstream release
    - New translations
    - Bugfixes
  * Note new translators in copyright file
  * Change Standards-Version to 3.9.3 (no changes)

 -- Christian M. Amsüss <chrysn@fsfe.org>  Sat, 02 Jun 2012 16:00:04 +0200

arandr (0.1.5-1) UNRELEASED; urgency=low

  * New upstream release
    - New translations
    - New `unxrandr` tool
  * Added a bug script that includes xrandr output in bug reports
  * Fixed typo in description (closes: #623518)
  * Note new translators in copyright file
  * Change Standards-Version to 3.9.2 (no changes)
  * Updated copyright file to DEP-5 release, now usable with config-model

 -- Christian M. Amsüss <chrysn@fsfe.org>  Sun, 09 Oct 2011 15:52:10 +0200

arandr (0.1.4-1) unstable; urgency=low

  * New upstream version
    - New translations
    - Bugfix
  * Note new translations in copyright file
  * Change Standards-Version to 3.9.1 (no changes)

 -- Christian M. Amsüss <chrysn@fsfe.org>  Mon, 13 Dec 2010 14:59:09 +0100

arandr (0.1.3-1) unstable; urgency=low

  * Change Standards-Version to 3.8.4 (no changes)
  * New upstream version
    - New translations
    - Minor bugfixes 
  * Add build dependency on python-docutils for man page generation
  * Note new translations in copyright file
  * Install new docs

 -- Christian M. Amsüss <chrysn@fsfe.org>  Fri, 12 Mar 2010 21:20:48 +0100

arandr (0.1.2-3) unstable; urgency=low

  * Add Vcs-* to debian/control
  * Switch to python-support (and simplify debian/rules; closes: #566037)
  * Switch to dpkg-source 3.0 (quilt) format

 -- Christian M. Amsüss <chrysn@fsfe.org>  Sun, 24 Jan 2010 22:05:38 +0100

arandr (0.1.2-2) unstable; urgency=low

  * Use --install-layout=deb (closes: #557790)
  * Copyright file according to DEP-5
  * Better use of debhelper
  * Updated to policy 3.8.3 (no changes required)

 -- Christian M. Amsüss <chrysn@fsfe.org>  Wed, 25 Nov 2009 18:41:38 +0100

arandr (0.1.2-1) unstable; urgency=low

  * New release (closes: #523903)
      + Accepts XRandR version 1.3
      + New command line switch to accept untested versions
  * Change Standards-Version to 3.8.1 (no changes)

 -- Christian M. Amsüss <chrysn@fsfe.org>  Mon, 13 Apr 2009 17:00:17 +0200

arandr (0.1.1-2) unstable; urgency=low

  * Fix GPL2/GPL3 mismatch

 -- Christian M. Amsüss <chrysn@fsfe.org>  Sun, 26 Oct 2008 11:26:45 +0100

arandr (0.1.1-1) unstable; urgency=low

  * Minor bugfixes, translation additions

 -- Christian M. Amsüss <chrysn@fsfe.org>  Mon, 21 Jul 2008 21:12:59 +0000

arandr (0.1-1) unstable; urgency=low

  * Initial release (closes: #484849)

 -- Christian M. Amsüss <chrysn@fsfe.org>  Thu, 5 Jun 2008 22:47:54 +0000

arandr (0.0-1) unstable; urgency=low

  * Initial pre-release package

 -- chrysn <chrysn@fsfe.org>  Mon, 17 Mar 2008 16:08:54 +0000



screenlayout/xrandr.py
# ARandR -- Another XRandR GUI
# Copyright (C) 2008 -- 2011 chrysn <chrysn@fsfe.org>
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""Wrapper around command line xrandr (mostly 1.2 per output features supported)"""

import os
import subprocess
import warnings

from .auxiliary import BetterList, Size, Position, Geometry, FileLoadError, FileSyntaxError, InadequateConfiguration, Rotation, ROTATIONS, NORMAL, NamedSize

import gettext
gettext.install('arandr')

SHELLSHEBANG='#!/bin/sh'

class Feature(object):
    PRIMARY = 1

class XRandR(object):
    DEFAULTTEMPLATE = [SHELLSHEBANG, '%(xrandr)s']

    def __init__(self, display=None, force_version=False):
        """Create proxy object and check for xrandr at `display`. Fail with
        untested versions unless `force_version` is True."""
        self.environ = dict(os.environ)
        if display:
            self.environ['DISPLAY'] = display

        version_output = self._output("--version")
        supported_versions = ["1.2", "1.3", "1.4", "1.5"]
        if not any(x in version_output for x in supported_versions) and not force_version:
            raise Exception("XRandR %s required."%"/".join(supported_versions))

        self.features = set()
        if not " 1.2" in version_output:
            self.features.add(Feature.PRIMARY)

    def _get_outputs(self):
        assert self.state.outputs.keys() == self.configuration.outputs.keys()
        return self.state.outputs.keys()
    outputs = property(_get_outputs)

    #################### calling xrandr ####################

    def _output(self, *args):
        p = subprocess.Popen(("xrandr",)+args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=self.environ)
        ret, err = p.communicate()
        status = p.wait()
        if status!=0:
            raise Exception("XRandR returned error code %d: %s"%(status,err))
        if err:
            warnings.warn("XRandR wrote to stderr, but did not report an error (Message was: %r)"%err)
        return ret

    def _run(self, *args):
        self._output(*args)

    #################### loading ####################

    def load_from_string(self, data):
        data = data.replace("%","%%")
        lines = data.split("\n")
        if lines[-1] == '': lines.pop() # don't create empty last line

        if lines[0] != SHELLSHEBANG:
            raise FileLoadError('Not a shell script.')

        xrandrlines = [i for i,l in enumerate(lines) if l.strip().startswith('xrandr ')]
        if len(xrandrlines)==0:
            raise FileLoadError('No recognized xrandr command in this shell script.')
        if len(xrandrlines)>1:
            raise FileLoadError('More than one xrandr line in this shell script.')
        self._load_from_commandlineargs(lines[xrandrlines[0]].strip())
        lines[xrandrlines[0]] = '%(xrandr)s'

        return lines

    def _load_from_commandlineargs(self, commandline):
        self.load_from_x()

        args = BetterList(commandline.split(" "))
        if args.pop(0) != 'xrandr':
            raise FileSyntaxError()
        options = dict((a[0], a[1:]) for a in args.split('--output') if a) # first part is empty, exclude empty parts

        for on,oa in options.items():
            o = self.configuration.outputs[on]
            os = self.state.outputs[on]
            o.primary = False
            if oa == ['--off']:
                o.active = False
            else:
                if '--primary' in oa:
                    if Feature.PRIMARY in self.features:
                        o.primary = True
                    oa.remove('--primary')
                if len(oa)%2 != 0:
                    raise FileSyntaxError()
                parts = [(oa[2*i],oa[2*i+1]) for i in range(len(oa)//2)]
                mode = ''
                rate = ''
                for p in parts:
                    if p[0] == '--mode':
                        mode = p[1]
                        if mode and rate:
                            for namedmode in os.modes:
                                if namedmode.name == mode + ' ' + rate + 'Hz':
                                    o.mode = namedmode
                                    break
                            else:
                                raise FileLoadError("Not a known mode: %s" % (mode + ' ' + rate + 'Hz'))
                    elif p[0] == '--rate':
                        rate = p[1]
                        if mode and rate:
                            for namedmode in os.modes:
                                if namedmode.name == mode + ' ' + rate + 'Hz':
                                    o.mode = namedmode
                                    break
                            else:
                                raise FileLoadError("Not a known mode: %s" % (mode + ' ' + rate + 'Hz'))
                    elif p[0] == '--pos':
                        o.position = Position(p[1])
                    elif p[0] == '--rotate':
                        if p[1] not in ROTATIONS:
                            raise FileSyntaxError()
                        o.rotation = Rotation(p[1])
                    else:
                        raise FileSyntaxError()
                o.active = True

    def load_from_x(self): # FIXME -- use a library
        self.configuration = self.Configuration(self)
        self.state = self.State()

        screenline, items = self._load_raw_lines()

        self._load_parse_screenline(screenline)

        for headline,details in items:
            if headline.startswith("  "): continue # a currently disconnected part of the screen i can't currently get any info out of
            if headline == "": continue # noise

            headline = headline.replace('unknown connection', 'unknown-connection')
            hsplit = headline.split(" ")
            o = self.state.Output(hsplit[0])
            assert hsplit[1] in ("connected","disconnected", 'unknown-connection')

            o.connected = (hsplit[1] in ('connected', 'unknown-connection'))

            primary = False
            if 'primary' in hsplit:
                if Feature.PRIMARY in self.features:
                    primary = True
                hsplit.remove('primary')

            if not hsplit[2].startswith("("):
                active = True

                geometry = Geometry(hsplit[2])

                modeid = hsplit[3].strip("()")

                if hsplit[4] in ROTATIONS: rotation = Rotation(hsplit[4])
                else: rotation = NORMAL
            else:
                active = False
                geometry = None
                modeid = None
                rotation = None

            o.rotations = set()
            for r in ROTATIONS:
                if r in headline:
                    o.rotations.add(r)

            currentname = None
            for d, w, h, f in details:
                n, m = d[0:2]
                n = n + f
                k = m.strip("()")
                try:
                    r = Size([int(w), int(h)])
                except ValueError:
                    raise Exception("Output %s parse error: modename %s modeid %s."%(o.name, n,k))
                if "*current" in d:
                    currentname = n
                for x in [ "+preferred", "*current" ]:
                    if x in d: d.remove(x)

                for old_mode in o.modes:
                    if old_mode.name == n:
                        if tuple(old_mode) != tuple(r):
                            warnings.warn("Supressing duplicate mode %s even though it has different resolutions (%s, %s)."%(n, r, old_mode))
                        break
                else:
                    # the mode is really new
                    o.modes.append(NamedSize(r, name=n))

            self.state.outputs[o.name] = o
            self.configuration.outputs[o.name] = self.configuration.OutputConfiguration(active, primary, geometry, rotation, currentname)

    def _load_raw_lines(self):
        output = self._output("--verbose")
        items = []
        screenline = None
        for l in output.split('\n'):
            if l.startswith("Screen "):
                assert screenline is None
                screenline = l
            elif l.startswith('\t'):
                continue
            elif l.startswith(2*' '): # [mode, width, height]
                l = l.strip()
                if l.startswith('h:'):
                    l = l[-len(l):l.index(" start")-len(l)]
                    items[-1][1][-1].append(l[l.rindex(' '):])
                elif l.startswith('v:'):
                    l1 = l[-len(l):l.index(" start")-len(l)]
                    items[-1][1][-1].append(l1[l1.rindex(' '):])
                    l1 = l[-len(l):l.index("Hz")-len(l)]
                    items[-1][1][-1].append(l1[l1.rindex(' '):]+'Hz')
                else: # mode
                    items[-1][1].append([l.split()])
            else:
                items.append([l, []])
        return screenline, items

    def _load_parse_screenline(self, screenline):
        assert screenline is not None
        ssplit = screenline.split(" ")

        ssplit_expect = ["Screen",None,"minimum",None,"x",None,"current",None,"x",None,"maximum",None,"x",None]
        assert all(a==b for (a,b) in zip(ssplit,ssplit_expect) if b is not None)

        self.state.virtual = self.state.Virtual(
                min = Size((int(ssplit[3]),int(ssplit[5][:-1]))),
                max = Size((int(ssplit[11]),int(ssplit[13])))
                )
        self.configuration.virtual = Size((int(ssplit[7]),int(ssplit[9][:-1])))

    #################### saving ####################

    def save_to_shellscript_string(self, template=None, additional=None):
        """Return a shellscript that will set the current configuration. Output can be parsed by load_from_string.

        You may specify a template, which must contain a %(xrandr)s parameter and optionally others, which will be filled from the additional dictionary."""
        if not template:
            template = self.DEFAULTTEMPLATE
        template = '\n'.join(template)+'\n'

        d = {'xrandr': "xrandr "+" ".join(self.configuration.commandlineargs())}
        if additional:
            d.update(additional)

        return template%d

    def save_to_x(self):
        self.check_configuration()
        self._run(*self.configuration.commandlineargs())

    def check_configuration(self):
        vmax = self.state.virtual.max

        for on in self.outputs:
            oc = self.configuration.outputs[on]
            #os = self.state.outputs[on]

            if not oc.active:
                continue

            # we trust users to know what they are doing (e.g. widget: will accept current mode, but not offer to change it lacking knowledge of alternatives)
            #if oc.rotation not in os.rotations:
            #    raise InadequateConfiguration("Rotation not allowed.")
            #if oc.mode not in os.modes:
            #    raise InadequateConfiguration("Mode not allowed.")

            x = oc.position[0] + oc.size[0]
            y = oc.position[1] + oc.size[1]

            if x > vmax[0] or y > vmax[1]:
                raise InadequateConfiguration(_("A part of an output is outside the virtual screen."))

            if oc.position[0] < 0 or oc.position[1] < 0:
                raise InadequateConfiguration(_("An output is outside the virtual screen."))

    #################### sub objects ####################

    class State(object):
        """Represents everything that can not be set by xrandr."""
        def __init__(self):
            self.outputs = {}

        def __repr__(self):
            return '<%s for %d Outputs, %d connected>'%(type(self).__name__, len(self.outputs), len([x for x in self.outputs.values() if x.connected]))

        class Virtual(object):
            def __init__(self, min, max):
                self.min = min
                self.max = max

        class Output(object):
            def __init__(self, name):
                self.name = name
                self.modes = []

            def __repr__(self):
                return '<%s %r (%d modes)>'%(type(self).__name__, self.name, len(self.modes))

    class Configuration(object):
        """Represents everything that can be set by xrandr (and is therefore subject to saving and loading from files)"""
        def __init__(self, xrandr):
            self.outputs = {}
            self._xrandr = xrandr

        def __repr__(self):
            return '<%s for %d Outputs, %d active>'%(type(self).__name__, len(self.outputs), len([x for x in self.outputs.values() if x.active]))

        def commandlineargs(self):
            args = []
            for on,o in self.outputs.items():
                args.append("--output")
                args.append(on)
                if not o.active:
                    args.append("--off")
                else:
                    if Feature.PRIMARY in self._xrandr.features:
                        if o.primary:
                            args.append("--primary")
                    modres=str(o.mode.name).split(" ")
                    args.append("--mode")
                    args.append(str(modres[0]))
                    args.append("--rate")
                    args.append(str(modres[1]).replace('Hz',''))
                    args.append("--pos")
                    args.append(str(o.position))
                    args.append("--rotate")
                    args.append(o.rotation)
            return args

        class OutputConfiguration(object):
            def __init__(self, active, primary, geometry, rotation, modename):
                self.active = active
                self.primary = primary
                if active:
                    self.position = geometry.position
                    self.rotation = rotation
                    if rotation.is_odd:
                        self.mode = NamedSize(Size(reversed(geometry.size)), name=modename)
                    else:
                        self.mode = NamedSize(geometry.size, name=modename)
            size = property(lambda self: NamedSize(Size(reversed(self.mode)), name=self.mode.name) if self.rotation.is_odd else self.mode)
